text/3371-cargo-target-dir-templates.md

```diff
@@ -0,0 +1,411 @@
+
+## Impact on `cargo ...` calls
+
+When calling `cargo` with a builtin call (e.g., `build`, `check` or `test`) where a templated `CARGO_TARGET_DIR` is active, `cargo` will first resolve the effective `CARGO_TARGET_DIR` and then proceed with the command as if `CARGO_TARGET_DIR` had been set directly. For third party tools (`cargo-*`), where cargo does not know about the relevant `Cargo.toml`, the tool will have to use [`cargo_metadata`](https://docs.rs/cargo_metadata), as is already expected today, to learn about the effective target directory.
+
+In the same vein, `cargo metadata` fills the target directory information with the absolute path and make no mention of the template in `CARGO_TARGET_DIR` since it can only be used with a single workspace at once.
+
+### `cargo clean`
+
+Currently, if `CARGO_TARGET_DIR` is set to anything but `target` for a project, `cargo clean` does not delete the `target/` directory if it exists, instead deleting the directory pointed by `CARGO_TARGET_DIR`. The same behavior is used for the templated version: if it set, `cargo clean` deletes `/path/to/<manifest-path-hash>/` and not `target/`.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+<!-- Why should we *not* do this? -->
+
+## One more option to find the target directory
+
+This introduces one more option to look at to find the target directory, which may complicate the life of external tools.
+
+This is mitigated by the forward link provided by default by `cargo` when using the templated form of `CARGO_TARGET_DIR`.
+
+## Hitting windows path length limits
+
+Depending on what naming scheme is used (e.g., a very long hash), we could hit the Windows path length limits if not careful.
+
+A mitigation for this is recommending a short prefix (in `CARGO_TARGET_DIR`) and using a hash that doesn't include that many characters but those are only mitigations and do not fully fix the underlying problem.
+
+## Transition period
+
+During the transition period, any `CARGO_TARGET_DIR` that was defined as containing `{manifest-path-hash}` will change meaning. `cargo`, for at least one stable version of Rust, should provide errors about this and point to either this RFC or its documentation to explain why the incompatiblity arised and how to fix it. In practice, paths with `{` or `}` in it are unlikely, even more with the exact key used by cargo here, so maybe no one will ever see the error, but it's better than silently breaking workflows.
+
+## Brace expansion
+
+Bash has [brace expansion](https://www.gnu.org/software/bash/manual/html_node/Brace-Expansion.html), other shells too. By using `{manifest-path-hash}` we risk users getting bitten by that behaviour. Brace expansion is not activated when there are no `,` nor `..` inside the `{}` so cargo should be fine. Since brace expansion is done at the shell level, cargo won't be able to detect it if it happens.
```

Comment:
```suggestion
Bash has [brace expansion](https://www.gnu.org/software/bash/manual/html_node/Brace-Expansion.html), other shells too. By using `{manifest-path-hash}` we risk users getting bitten by that behaviour. Brace expansion is only activated when there are `,` nor `..` inside the `{}` so cargo should be fine. Since brace expansion is done at the shell level, cargo won't be able to detect it if it happens.
```
Removed the double negative to make it easier to parse

tip:  [semantic line feeds](https://rhodesmill.org/brandon/2012/one-sentence-per-line/) make diffing markdown documents easier.

Reply:
Done in https://github.com/rust-lang/rfcs/commit/3a6fa3e6848a295f7544bf6ac430119b8c1ffc94